File: src/1_100/0024_Swap_Nodes_in_Pairs/Swap_Nodes_in_Pairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
	
	//另一种实现方式，基于栈，遍历的时候往栈中push两个元素，再弹出两个元素，就可以达到翻转的效果
	public ListNode swapPairs_2(ListNode head) {
		if(head==null || head.next==null) {
			return head;
		}
		ListNode dummy = new ListNode(-1);
		ListNode p = head;
		head = dummy;
		Stack<ListNode> stack = new Stack<ListNode>();
		while(p!=null && p.next!=null) {
			stack.add(p);
			stack.add(p.next);
			p = p.next.next;
			dummy.next = stack.pop();
			dummy = dummy.next;
			dummy.next = stack.pop();
			dummy = dummy.next;
		}
		if(p != null) {
			dummy.next = p;
		} else {
			dummy.next = null;
		}
		//dump(head,"head");
		return head.next;
	}
	
    public ListNode swapPairs(ListNode head) {
        if(head==null || head.next==null) {
            return head;
        }
        ListNode saveToRes = new ListNode(-1);
        saveToRes.next = head.next;
        ListNode p = new ListNode(-1);
        p.next = head;
        ListNode a = null;
        ListNode b = null;
        ListNode tmp = null;
        while(p.next!=null && p.next.next!=null) {
            a = p.next;
            b = p.next.next;
            tmp = b.next;
            b.next = a;
            a.next = tmp;
            p.next = b;
            p = a;
        }
        //System.out.println("??????");
        return saveToRes.next;
    }
}
            

File: src/1_100/0024_Swap_Nodes_in_Pairs/Swap_Nodes_in_Pairs.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if(head==None or head.next==None):
            return head
        next_node = head.next
        head.next = self.swapPairs(next_node.next)
        next_node.next = head
        return next_node
		
		
	# 不借助stack的实现方式	
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        p = ListNode(-1)
        a,b,tmp,p.next = p,p,p,head
        while b and b.next and b.next.next:
            a,b,tmp.next = a.next,b.next.next,b.next.next
            a.next,b.next,tmp,b = b.next,a,a,a
        return p.next
		
		
	# 借助stack的方式，精简版
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        dummy = ListNode(-1)
        p,head,stack = head,dummy,[]
        while p and p.next:
            _,_,p = stack.append(p),stack.append(p.next),p.next.next
            dummy.next,dummy.next.next = stack.pop(),stack.pop()
            dummy,dummy.next = dummy.next.next,None
        if p:
            dummy.next = p
        return head.next
		
		
		
		

		